backend/azureblob/azureblob.go

```diff
@@ -718,19 +719,32 @@ func NewFs(ctx context.Context, name, root string, m configmap.Mapper) (fs.Fs, e
 		} else {
 			serviceURL = azblob.NewServiceURL(*u, pipeline)
 		}
-	case opt.ServicePrincipalFile != "":
+	case opt.ServicePrincipalFile != "" || opt.UseEnvCreds:
 		// Create a standard URL.
 		u, err = url.Parse(fmt.Sprintf("https://%s.%s", opt.Account, opt.Endpoint))
 		if err != nil {
 			return nil, fmt.Errorf("failed to make azure storage url from account and endpoint: %w", err)
 		}
-		// Try loading service principal credentials from file.
-		loadedCreds, err := os.ReadFile(env.ShellExpand(opt.ServicePrincipalFile))
-		if err != nil {
-			return nil, fmt.Errorf("error opening service principal credentials file: %w", err)
+		var spCredentials servicePrincipalCredentials
+		if opt.ServicePrincipalFile != "" {
+			// Try loading service principal credentials from file.
+			loadedCreds, err := os.ReadFile(env.ShellExpand(opt.ServicePrincipalFile))
+			if err != nil {
+				return nil, fmt.Errorf("error opening service principal credentials file: %w", err)
+			}
+
+			if err := json.Unmarshal(loadedCreds, &spCredentials); err != nil {
+				return nil, fmt.Errorf("error parsing credentials from JSON file: %w", err)
+			}
+		} else {
+			spCredentials = servicePrincipalCredentials {
```

Comment:
What happens if any of these env vars are missing? Should we check a minimum set is present?

Reply:
All three must be present.  Right now, the behavior in the presence of a missing environment variable would be identical to having a JSON file that omits that value.  The test suite gives an example:
https://github.com/rclone/rclone/blob/591fc3609aea47441fbd4bc01235291efea9bb4e/backend/azureblob/azureblob_test.go#L62

That may not be the best error message for this case (nor, perhaps, for the case under test), and so I am open to suggestions.

Reply:
I think that is OK then.